Translate to/from specs v2.4 and v3.1

LorbusChris opened the pull request:
The Ignition config specs v2.4 and v.3.1 are the currently
stable ones in Ignition 0.x and 2.x respectively.

bgilbert commented:
We should never output an experimental spec.  There's no guarantee of compatibility even within an experimental version, and once an experimental spec is stabilized, the `-experimental` variant is no longer accepted by Ignition.

It might make sense to bump to 2.4 and 3.1, but we should be careful.  If the converter is being used to provision OS versions which might not support those specs, we'll cause provisioning failures.

bgilbert commented on internal/main.go:
```diff
@@ -94,7 +94,7 @@ func main() {
 	if downtranslate {
 		// translate from 3 to 2
-		cfg, rpt, err := v3_0.Parse(dataIn)
+		cfg, rpt, err := v3.Parse(dataIn)
```
Ignition 2.x, unlike 0.x, doesn't automatically chain to the previous spec version if it can't parse with the specified one.  So we need to do that ourselves here, or else we won't be able to parse 3.0 specs.

LorbusChris replied:
Fixed :)

LorbusChris commented:
I've added functionality to translate the new fields in 2.4/3.1.
This changes the main function to translate up to 3.1 and down to 2.2

Also added more individual functions for translating from/to specific versions (they might not all be required - what I'd like to use for MCO is capability to translate back and forth between 2.2 and 3.1).

LorbusChris commented on translate3_0to2_2.go:
```diff
@@ -320,7 +321,12 @@ func translateFiles3to2(files []types.File, fss []string) (ret []old.File) {
 		// we should omit it if append is specified as we want the "unset"
 		// default.
-		if f.Node.Overwrite == nil && len(f.FileEmbedded1.Append) == 0 {
+		if f.Node.Overwrite != nil {
+			file.Node.Overwrite = f.Node.Overwrite
+		}
+
+		// If append is true overwrite must be false in spec 3
+		if f.FileEmbedded1.Append != nil {
```
@bgilbert note this change in logic here (I'll split that into a separate commit)

LorbusChris replied:
this is in a separate commit now.

LorbusChris commented:
@bgilbert this is ready for review :)

yuqi-zhang commented:
The code itself seems fine, however with this PR, we would have:
 - 2.3 -> 3.0 (original)
 - 2.4 -> 3.1 (latest versions)
 - 3.0 -> 2.2 (original)
 - 3.0 -> 2.3 (not sure why we need this, I assume for compatibility?)
 - 3.1 -> 2.2 (not sure why we need this, do we want to start the MCO on 3.1 for master branch? The FCOS branch today is on 3.0)
 - 3.1 -> 2.4 (latest versions)

It feels a bit redundant. Do we need all that compatibility? The FCOS MCO branch today is on 3.0. Are we able to convert between 3.0 to 3.1? If so I think we can drop some of the excessive translations.

LorbusChris commented:
yep, the intention is to go straight to spec 3.1 in OCP (I.e. https://github.com/openshift/machine-config-operator/pull/1778, after dual support has propagated through to MCD on RHCOS)

I think I liked the idea of adding 3.0 -> 2.3 so we can translate back and forth between them, but it's not really necessary indeed. I'll remove that one.

yuqi-zhang commented:
Right, hopefully after OCP stabilizes and is able to remove spec 2.x entirely, we would no longer need to support translations (unless some other major case shows up). So we won't have to add more backward compatible versions in the future. Does that make sense?

LorbusChris commented:
Dropped the 3.0 -> 2.3 conversion func

bgilbert commented on internal/main.go:
```diff
@@ -93,14 +96,23 @@ func main() {
 		}
 
-		newCfg, err := ignconverter.Translate3_0to2_2(cfg)
+		newCfg, err := ignconverter.Translate3_1to2_2(cfg)
```
How is the command-line tool expected to be used?  Would it make sense to translate to 2.4, since that conversion is not as lossy?

LorbusChris replied:
agreed. Changed it accordingly

LorbusChris commented:
@bgilbert I believe I've addressed all comments. PTAL

LorbusChris commented on translate3_0to2_2.go:
```diff
@@ -318,10 +318,14 @@ func translateFiles3_0to2_2(files []types.File, fss []string) (ret []old.File) {
+		// Overwrite defaults to false in spec 3 and true in spec 2.
+		// We want the "unset" default.
+		if f.Node.Overwrite != nil {
+			file.Node.Overwrite = f.Node.Overwrite
```
I've removed this

LorbusChris commented on util.go:
```diff
@@ -14,6 +14,56 @@
+// Error definitions
+
+// ErrUsesNetworkd is the error for including networkd configs
+var ErrUsesNetworkd = errors.New("config includes deprecated networkd section - use Files instead")
```
switched back to using `Error` suffix everywhere (somehow my linter doesn't like it, though, and wants it to be an `Err` prefix)

bgilbert commented on pkg/translate23to30/translate23to30.go:
```diff
@@ -368,7 +368,7 @@ func translateFiles(files []old.File, m map[string]string) (ret []types.File) {
 		// i.e. spec 2 files with append true must be translated to spec 3 files with overwrite false
 		if f.FileEmbedded1.Append == true {
-			f.Node.Overwrite = boolPStrict(false)
+			f.Node.Overwrite = util.BoolPStrict(false)
```
Fixup targeted to wrong commit?

bgilbert commented on pkg/translate30to22/translate30to22.go:
```diff
@@ -349,7 +349,7 @@ func translateFiles(files []types.File, fss []string) (ret []old.File) {
 				appendFile.FileEmbedded1.Append = true
 				// In spec 2, overwrite must be false on appended files
-				appendFile.Node.Overwrite = boolPStrict(false)
+				appendFile.Node.Overwrite = util.BoolPStrict(false)
```
Fixup targeted to wrong commit?

bgilbert commented on translate_test.go:
```diff
@@ -450,7 +621,7 @@ var (
 						Filesystem: "root",
 						Path:       "/etc/motd",
-						Overwrite:  boolPStrict(false),
+						Overwrite:  util.BoolPStrict(false),
```
Fixup targeted to wrong commit?

LorbusChris replied:
this one is actually correct I think

bgilbert commented on translate_test.go:
```diff
@@ -299,7 +470,7 @@ var (
 					Node: types3_0.Node{
 						Path:      "/empty",
-						Overwrite: util.BoolPStrict(true),
+						Overwrite: util.BoolPStrict(false),
```
Fix targeted to wrong commit?

bgilbert commented on pkg/translate31to22/translate31to22.go:
```diff
@@ -0,0 +1,428 @@
+	"strconv"
+	"strings"
+
+	"github.com/coreos/ign-converter/pkg/util"
```
Appears in an inconsistent place between 30to22 and 31to22, and so makes diff noisier.

LorbusChris replied:
fixed

bgilbert commented on pkg/translate24to31/translate24to31.go:
```diff
@@ -0,0 +1,443 @@
+		path := filepath.Join("/", fsMap[file.Filesystem], file.Path)
+		name := fmt.Sprintf("File: %s", path)
+		if duplicate, isDup := entryMap[path]; isDup {
+			return util.DuplicateInodeError{duplicate, name}
```
23to30 returns a pointer to this, noisifying the diff.

LorbusChris replied:
I don't follow. Which part to you mean specifically?

bgilbert replied:
Sorry, I put this comment on the wrong line.  I was referring to https://github.com/coreos/ign-converter/blob/8bfca335dd8aeaf28b5dec8fd6f4cf87bfe06786/translate/v23tov30/v23tov30.go#L99.

bgilbert commented on pkg/translate24to31/translate24to31.go:
```diff
@@ -0,0 +1,443 @@
+			f.Node.Overwrite = util.BoolP(true)
+		}
+		// In spec 3, overwrite must be false if source is unspecified or append is true
+		if f.Contents.Source == "" || f.FileEmbedded1.Append == true {
```
Wasn't updated in 24to31.

LorbusChris commented on internal/main.go:
```diff
@@ -22,10 +22,14 @@ import (
+	"github.com/coreos/ignition/v2/config/v3_1"
+	translateTo3_1 "github.com/coreos/ignition/v2/config/v3_1/translate"
+
+	"github.com/coreos/ign-converter/pkg/translate24to31"
```
go-lint isn't happy with underscores in package names. I could do `.../ign-converter/translate/v24tov31` if you prefer.

bgilbert replied:
SGTM.

LorbusChris commented on pkg/translate30to22/translate30to22.go:
```diff
@@ -340,35 +343,37 @@ func translateFiles3to2(files []types.File, fss []string) (ret []old.File) {
 				}
 				appendFile.FileEmbedded1.Contents.Verification.Hash = fc.Verification.Hash
 				appendFile.FileEmbedded1.Append = true
+				// In spec 2, overwrite must be false on appended files
```
I've added more commentary. Does it suffice?

bgilbert replied:
Sure, that works.

bgilbert replied:
Remember to update v31tov22 also.

LorbusChris commented:
@bgilbert I've pushed the next iteration of this :)
